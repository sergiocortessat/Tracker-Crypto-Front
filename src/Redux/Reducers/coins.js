const initialState = {
  coins: [],
};

const coinReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'ALL_COINS':
      return {
        ...state,
        coins: action.payload,
      };
    default:
      return state;
  }
};

export default coinReducer;
