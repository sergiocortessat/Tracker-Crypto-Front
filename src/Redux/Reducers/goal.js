const initialState = {
  goals: [],
};

const goalReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'ALL_GOALS':
      return {
        ...state,
        goal: action.payload,
      };
    default:
      return state;
  }
};

export default goalReducer;
