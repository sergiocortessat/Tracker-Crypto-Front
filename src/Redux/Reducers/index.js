import { combineReducers } from 'redux';
import Coins from './coins';
import User from './user';

const reducers = combineReducers({
  coins: Coins,
  user: User,
});

export default reducers;
