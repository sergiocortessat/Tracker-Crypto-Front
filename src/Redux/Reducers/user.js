const initialState = {
  user: [],
};

const userReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'CURRENT_USER':
      return {
        ...state,
        user: action.payload,
      };
    default:
      return state;
  }
};

export default userReducer;
