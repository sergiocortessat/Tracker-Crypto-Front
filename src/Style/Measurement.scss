$grey: #67899c;
$green: #addc91;

.crypto-container {
  height: 100%;
  color: $grey;
  display: flex;
  flex-direction: column;
  justify-content: space-between;

  .crypto-name {
    text-align: center;
    margin-top: 10px;
    font-size: 2rem;
  }

  form {
    text-align: center;
    margin: 30px 0 30px 0;

    input {
      width: 60px;
      text-align: center;
    }
  }

  .measurement {
    background: lightgrey;
    flex: 1;
    // min-height: 250px;

    h2 {
      font-size: 1.6rem;
      padding: 20px 0 10px 20px;
    }

    .measurement-list {
      margin-top: 10px;
      display: flex;
      flex-direction: column;
      align-items: center;
      color: #67899c;

      .measurement-item {
        display: flex;
        flex-direction: row;
        width: 90%;
        margin-bottom: 10px;
        background-color: white;
        height: 50px;
        align-items: center;
        justify-content: space-evenly;
        border-radius: 10px;

        .measurement-progress {
          width: 60px;
          padding: 15px 8px 15px 8px;
        }

        .item {
          display: flex;
          align-items: center;
          width: 100%;
          color: $grey;
        }

        p {
          margin: 0 0 0 2px;
          font-size: 1rem;
          width: 90%;
          color: black;
        }

        time {
          text-align: center;
          margin: 0 auto;
          width: 90%;
          font-size: 0.6rem;
          color: black;
        }

        button {
          background-color: $green;
          text-align: center;
          margin: 0 15px;
          padding: 5px;
          border-radius: 15px;
          color: white;
        }
      }
    }
  }
}
