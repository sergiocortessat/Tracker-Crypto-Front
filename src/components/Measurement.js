import { useAuth0 } from '@auth0/auth0-react';
import React, { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import Moment from 'react-moment';
import moment from 'moment';
import { deleteMeasurement, getGoal } from '../API/API';
import MainCircularProgress from './MainCirculaProgress';
import SetGoals from './setGoals';
import '../Style/ProgressBar.scss';
import '../Style/Measurement.scss';
import customAlert from './PopUpAlert';

const Measurement = () => {
  const { user, getAccessTokenSilently } = useAuth0();
  const { id } = useParams();
  const [measures, setMeasures] = useState([]);
  const [goal, setGoal] = useState(null);
  const [percentage, setPercentage] = useState(0);
  const [today, setToday] = useState([]);
  const [yesterday, setYesterday] = useState([]);
  const [lastWeek, setLastWeek] = useState([]);
  const now = moment(new Date());
  const coinName = ['Bitcoin', 'Ethereum', 'Cardano', 'Uniswap', 'XRP', 'Polkadot'];
  const [count, setCount] = useState(0);

  useEffect(() => {
    getGoal(user.sub).then((res) => {
      const temp = res.filter((m) => m.coin_id === Number(id))[0];
      setGoal(temp.goal);
      setMeasures(temp.measurements);
    });
  }, [count]);

  useEffect(() => {
    const result = measures.reduce((acc, cur) => acc + cur.unit, 0);
    const percentage = ((result * 100) / goal);
    setPercentage(percentage);
  }, [measures, goal]);

  const handleDelete = (id) => {
    customAlert('Deleted');
    getAccessTokenSilently().then((accesToken) => {
      deleteMeasurement(id, accesToken);
    });
    setTimeout(() => {
      setCount(count + 1);
    }, 500);
  };

  const fetchData = async () => {
    try {
      setToday(
        measures.filter(
          (item) => Number(
            moment
              .duration(now.diff(moment(item.created_at).format('L')))
              .asDays(),
          ) < 1,
        ),
      );
      setYesterday(
        measures.filter(
          (item) => Math.floor(Number(
            moment
              .duration(now.diff(moment(item.created_at).format('L')))
              .asDays(),
          )) >= 1
              && Math.floor(Number(
                moment
                  .duration(now.diff(moment(item.created_at).format('L')))
                  .asDays(),
              )) < 2,
        ),
      );
      setLastWeek(
        measures.filter(
          (item) => Math.floor(Number(
            moment.duration(now.diff(item.created_at)).asDays(),
          )) > 2,
        ),
      );
    } catch (e) {
      return e;
    }
    return null;
  };

  useEffect(() => {
    fetchData();
  }, [measures]);

  const individualPercentage = (unit) => {
    const percentage = ((unit * 100) / goal);
    return percentage;
  };

  return (
    <div className="crypto-container">
      <div className="crypto-name">
        {coinName[Number(id) - 1]}
      </div>

      <div className="progress-icon-measurement">
        <MainCircularProgress percentage={percentage} />
      </div>

      <SetGoals coin={Number(id)} setChange={setGoal} />

      <div className="measurement">
        {today.length > 0 && (<h2>Today</h2>)}
        <div className="measurement-list">
          {today && today.map((measures) => (
            <div key={measures.id} className="measurement-item">
              <div className="measurement-progress">
                <MainCircularProgress percentage={individualPercentage(measures.unit)} />
              </div>
              <div className="item">
                <p>
                  Units:
                  {' '}
                  {measures.unit}
                </p>
                <Moment fromNow>{measures.created_at}</Moment>
                <button type="button" onClick={() => handleDelete(measures.id)}>Delete</button>
              </div>
            </div>
          ))}
        </div>

        {yesterday.length > 0 && (<h2>Yesterday</h2>)}
        <div className="measurement-list">
          {yesterday && yesterday.map((measures) => (
            <div key={measures.id} className="measurement-item">
              <div className="measurement-progress">
                <MainCircularProgress percentage={individualPercentage(measures.unit)} />
              </div>
              <div className="item">
                <p>
                  Units:
                  {' '}
                  {measures.unit}
                </p>
                <Moment fromNow>{measures.created_at}</Moment>
                <button type="button" onClick={() => handleDelete(measures.id)}>Delete</button>
              </div>
            </div>
          ))}
        </div>
        {lastWeek.length > 0 && (<h2>Last Week</h2>)}
        <div className="measurement-list">
          {lastWeek && lastWeek.map((measures) => (
            <div key={measures.id} className="measurement-item">
              <div className="measurement-progress">
                <MainCircularProgress percentage={individualPercentage(measures.unit)} />
              </div>
              <div className="item">
                <p>
                  Units:
                  {' '}
                  {measures.unit}
                </p>
                <Moment fromNow>{measures.created_at}</Moment>
                <button type="button" onClick={() => handleDelete(measures.id)}>Delete</button>
              </div>
            </div>
          ))}
        </div>
      </div>
    </div>
  );
};

export default Measurement;
