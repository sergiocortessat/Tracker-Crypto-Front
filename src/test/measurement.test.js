import { Provider } from 'react-redux';
import { createStore } from 'redux';
import renderer from 'react-test-renderer';
import {
  BrowserRouter,
} from 'react-router-dom';
import Reducers from '../Redux/Reducers';
import Measurement from '../components/Measurement';

const store = createStore(Reducers);

it('renders correctly', () => {
  const addMeasurement = renderer
    .create(
      <Provider store={store}>
        <BrowserRouter>
          <Measurement />
        </BrowserRouter>
      </Provider>,
    )
    .toJSON();
  expect(addMeasurement).toMatchSnapshot();
});
