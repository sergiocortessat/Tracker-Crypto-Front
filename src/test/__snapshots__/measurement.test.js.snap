// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="crypto-container"
>
  <div
    className="crypto-name"
  />
  <div
    className="progress-icon-measurement"
  >
    <div
      data-test-id="CircularProgressbarWithChildren"
    >
      <div
        style={
          Object {
            "height": "100%",
            "position": "relative",
            "width": "100%",
          }
        }
      >
        <svg
          className="CircularProgressbar "
          data-test-id="CircularProgressbar"
          style={Object {}}
          viewBox="0 0 100 100"
        >
          <path
            className="CircularProgressbar-trail"
            d="
      M 50,50
      m 0,-45
      a 45,45 0 1 1 0,90
      a 45,45 0 1 1 0,-90
    "
            fillOpacity={0}
            strokeWidth={10}
            style={
              Object {
                "strokeDasharray": "282.7433388230814px 282.7433388230814px",
                "strokeDashoffset": "0px",
                "strokeLinecap": "butt",
              }
            }
          />
          <path
            className="CircularProgressbar-path"
            d="
      M 50,50
      m 0,-45
      a 45,45 0 1 1 0,90
      a 45,45 0 1 1 0,-90
    "
            fillOpacity={0}
            strokeWidth={10}
            style={
              Object {
                "strokeDasharray": "282.7433388230814px 282.7433388230814px",
                "strokeDashoffset": "282.7433388230814px",
                "strokeLinecap": "butt",
              }
            }
          />
          <text
            className="CircularProgressbar-text"
            style={Object {}}
            x={50}
            y={50}
          >
            0
          </text>
        </svg>
        <div
          data-test-id="CircularProgressbarWithChildren__children"
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "column",
              "height": "100%",
              "justifyContent": "center",
              "marginTop": "-100%",
              "position": "absolute",
              "width": "100%",
            }
          }
        >
          <div
            style={
              Object {
                "height": "100%",
                "position": "absolute",
                "transform": "rotate(0turn)",
              }
            }
          >
            <div
              style={
                Object {
                  "background": "#fff",
                  "height": "20%",
                  "width": "2px",
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "height": "100%",
                "position": "absolute",
                "transform": "rotate(0.08333333333333333turn)",
              }
            }
          >
            <div
              style={
                Object {
                  "background": "#fff",
                  "height": "20%",
                  "width": "2px",
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "height": "100%",
                "position": "absolute",
                "transform": "rotate(0.16666666666666666turn)",
              }
            }
          >
            <div
              style={
                Object {
                  "background": "#fff",
                  "height": "20%",
                  "width": "2px",
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "height": "100%",
                "position": "absolute",
                "transform": "rotate(0.25turn)",
              }
            }
          >
            <div
              style={
                Object {
                  "background": "#fff",
                  "height": "20%",
                  "width": "2px",
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "height": "100%",
                "position": "absolute",
                "transform": "rotate(0.3333333333333333turn)",
              }
            }
          >
            <div
              style={
                Object {
                  "background": "#fff",
                  "height": "20%",
                  "width": "2px",
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "height": "100%",
                "position": "absolute",
                "transform": "rotate(0.41666666666666663turn)",
              }
            }
          >
            <div
              style={
                Object {
                  "background": "#fff",
                  "height": "20%",
                  "width": "2px",
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "height": "100%",
                "position": "absolute",
                "transform": "rotate(0.5turn)",
              }
            }
          >
            <div
              style={
                Object {
                  "background": "#fff",
                  "height": "20%",
                  "width": "2px",
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "height": "100%",
                "position": "absolute",
                "transform": "rotate(0.5833333333333333turn)",
              }
            }
          >
            <div
              style={
                Object {
                  "background": "#fff",
                  "height": "20%",
                  "width": "2px",
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "height": "100%",
                "position": "absolute",
                "transform": "rotate(0.6666666666666666turn)",
              }
            }
          >
            <div
              style={
                Object {
                  "background": "#fff",
                  "height": "20%",
                  "width": "2px",
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "height": "100%",
                "position": "absolute",
                "transform": "rotate(0.75turn)",
              }
            }
          >
            <div
              style={
                Object {
                  "background": "#fff",
                  "height": "20%",
                  "width": "2px",
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "height": "100%",
                "position": "absolute",
                "transform": "rotate(0.8333333333333333turn)",
              }
            }
          >
            <div
              style={
                Object {
                  "background": "#fff",
                  "height": "20%",
                  "width": "2px",
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "height": "100%",
                "position": "absolute",
                "transform": "rotate(0.9166666666666666turn)",
              }
            }
          >
            <div
              style={
                Object {
                  "background": "#fff",
                  "height": "20%",
                  "width": "2px",
                }
              }
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <form
    action="/action_page.php"
  >
    <label
      htmlFor="quantity"
    >
      Current goal
       
      <input
        id={NaN}
        inputMode="numeric"
        max="1000"
        min="1"
        name="quantity"
        onChange={[Function]}
        placeholder=""
        type="number"
        value=""
      />
    </label>
  </form>
  <div
    className="measurement"
  >
    <div
      className="measurement-list"
    />
    <div
      className="measurement-list"
    />
    <div
      className="measurement-list"
    />
  </div>
</div>
`;
